### PR TITLE
Diagnostic provider for '!pip install' and `!conda install'

### PR DESCRIPTION
**Pull request by DonJayamanne**
For #6864

### CODE REVIEW

File: src/client/datascience/notebook/diagnosticsProvider.ts [hunk 1]
```diff
@@ -0,0 +1,234 @@
...
+        // Process just the first 100 cells to avoid blocking the UI.
+        notebook.getCells().forEach((cell, i) => (i < 100 ? this.queueCellForProcessing(cell) : undefined));
+    }
+
+    private queueCellForProcessing(cell: NotebookCell): void {
+        this.cellsToProces.add(cell);
+        this.analyzeNotebookCells();
+    }
+    private analyzeNotebookCells() {
+        if (this.cellsToProces.size === 0) {
```
Comment (rchiodo) [inline src/client/datascience/notebook/diagnosticsProvider.ts:175]: Spelling 'cellsToProces' should be 'cellsToProcess'

File: src/client/datascience/notebook/diagnosticsProvider.ts [hunk 2]
```diff
@@ -0,0 +1,234 @@
...
+        this.problems.delete(cell.document.uri);
+        const cellsUrisWithProblems = this.notebooksProcessed.get(cell.notebook) || new Map<CellUri, CellVersion>();
+        cellsUrisWithProblems.set(cell.document.uri.toString(), cell.document.version);
+        this.notebooksProcessed.set(cell.notebook, cellsUrisWithProblems);
+
+        // For perf reasons, process just the first 50 lines.
+        for (let i = 0; i < Math.min(cell.document.lineCount, 50); i++) {
+            const line = cell.document.lineAt(i);
+            const text = line.text;
+            if (text.trim().startsWith('!pip install')) {
```
Comment (rchiodo) [inline src/client/datascience/notebook/diagnosticsProvider.ts:207]: These can also be inside an if condition so you might want to use a regex instead.

Comment (rchiodo) [reply src/client/datascience/notebook/diagnosticsProvider.ts:207]: Never mind, missed the trim.

**Review by rchiodo — APPROVED** [review]

File: src/client/datascience/notebook/diagnosticsProvider.ts (hunk 2 above)
Comment (DonJayamanne) [reply src/client/datascience/notebook/diagnosticsProvider.ts:207]: Even if inside if condition, this would be in a separate line

**Pull request merged**